NetP Invite code screen PR 2

graeme opened the pull request:
Task/Issue URL: https://app.asana.com/0/0/1205084446087083/f
BSK PR: https://github.com/duckduckgo/BrowserServicesKit/pull/441
CC:

**Description:**

Second PR for the NetP Invite Code flow. You can find designs in the linked task above. This builds upon #1881 and adds to the existing Invite Code presentation logic in the `NetworkProtectionInviteViewModel`. It also fleshes out and designs the `NetworkProtectionInviteView` which has two states:

- codeEntry
- success

This also adds some tests which required some mocks / test helpers for types in the NetworkProtection library in BSK. As we will likely need these on macOS when we come to add more tests there, I added a NetworkProtectionTestUtils library to BSK. This PR points to that BSK branch so I will make sure to point to the BSK release once that’s approved. BSK PR linked above.

**Steps to test this PR:**

1. Check out this branch
2. Run the app
3. Ensure you’re authenticated as internal
4. Navigate to the Settings
5. Select Network Protection
6. If you’ve already enabled it, tap the button to clear the invite code then navigate back and select the NetP item again
7. Enter a random string and submit it to trigger the error label.
8. Now enter an invite code from [Network Protection Invite Codes](https://app.asana.com/0/1203135672790568/1204303483337129/f) and submit it
9. The success screen will be shown. Click Get Started to launch the status view

**Copy Testing**:

* [ ] Use of correct apostrophes in new copy, ie `’` rather than `'`

**Orientation Testing**:

* [x] Portrait
* [x] Landscape

**Device Testing**:

* [ ] iPhone SE (1st Gen)
* [ ] iPhone 8
* [ ] iPhone X
* [x] iPhone 14 Pro
* [ ] iPad

**OS Testing**:

* [x] iOS 14
* [ ] iOS 15
* [x] iOS 16

**Theme Testing**:

* [x] Light theme
* [x] Dark theme

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)

graeme committed:
Add convenience init for token store

graeme committed:
Add a root view to choose which netp view to show

graeme committed:
Add invite lock assets

graeme committed:
Implement initial invite view

graeme committed:
Create empty invite view

graeme committed:
Show root and therefore invite view from settings

graeme committed:
Add convenience init for redemption coordinator

graeme committed:
Copy over some useful invite code from macOS

graeme committed:
Remove invite entry from Status View

graeme committed:
Add some mocks and test helpers

To be moved to BSK

graeme committed:
Write tests + complete invite presentation logic

graeme committed:
Add success view

graeme committed:
Replace delegation with completion handler

graeme committed:
Show an alert on error

graeme committed:
Move ClearTextField out of Autofill

graeme committed:
Disable submit on empty string

graeme committed:
Add submit button text

graeme committed:
and the text

graeme committed:
General tidy up

graeme committed:
Pop 'n' push the root view on get started

graeme committed:
Move mocks to BSK

graeme committed:
Make image correct size

graeme committed:
Reduce padding below message text

graeme committed:
Create NetPRootViewController for dark styling

graeme committed:
Fix invite view dark text

graeme committed:
Remove unused colors

graeme committed:
Add a nav title

graeme committed:
Disable failing invite code error test

graeme commented on DuckDuckGo/AutofillLoginDetailsView.swift:
```diff
@@ -236,43 +236,6 @@ struct AutofillLoginDetailsView: View {
     }
 }
 
-struct ClearTextField: View {
```
This makes more sense in the shared DuckUI library

diegoreymendez replied:
This confused me a bit because it was ultimately moved to a different file right next to this one.

That's fine, but just adding a comment to make sure that was the intention.

graeme replied:
My comment was incorrect, it was actually moved into its own file in the shared UI components folder, not library. DuckUI doesn’t seem to have much in it at all. Nonetheless, I thought it didn’t make much sense in the AutofillLoginDetailsView file when it’s now being used from a couple of places.

graeme commented on DuckDuckGo/NetworkProtectionInviteViewModel.swift:
```diff
@@ -19,67 +19,72 @@
-    private static let errorEvents: EventMapping<NetworkProtectionError> = .init { _, _, _, _ in
-
-    }
+    private let completion: () -> Void
```
Used to pop the nav stack and navigate to the status view

graeme commented on DuckDuckGo/NetworkProtectionInviteViewModel.swift:
```diff
@@ -19,67 +19,72 @@
+        completion()
     }
 
+    // MARK: Dev only. Will be removed during https://app.asana.com/0/0/1205084446087078/f
```
I was using this same ViewModel for the basic invite code entry on the status view. Rather than move it in there, I will keep these functions here for now until I’ve added a replacement, probably in the debug menu, as part of the status view subtask.

graeme commented on DuckDuckGo/NetworkProtectionRootView.swift:
```diff
@@ -20,25 +20,32 @@
+    let inviteCompletion: () -> Void
 
     var body: some View {
+        let inviteViewModel = NetworkProtectionInviteViewModel(
```
This is not idea for now. It would be good to use a more elegant dependency injection mechanism than letting the View do it. But for now, this will do. Once the rest of the feature is more fleshed out, I’ll tidy this up.

diegoreymendez replied:
What's a "DI mechanism"?

graeme replied:
dependency injection (updated original comment)

graeme commented on DuckDuckGo/NetworkProtectionRootViewController.swift:
```diff
@@ -0,0 +1,54 @@
+
+import SwiftUI
+
+final class NetworkProtectionRootViewController: UIHostingController<NetworkProtectionRootView> {
```
This subclass is needed so we can apply some styling to background of the root view. This will probably be a better place to manage the dependency injection in the future too.

graeme commented on DuckDuckGo/NetworkProtectionStatusView.swift:
```diff
@@ -64,38 +64,22 @@ struct NetworkProtectionStatusView: View {
     @ViewBuilder
     func inviteCodeEntry() -> some View {
         Section {
-            VStack(alignment: .leading) {
```
Gets rid of the initial dev-designed invite entry on the dev-designed status view

graeme commented on DuckDuckGo/SettingsViewController.swift:
```diff
@@ -457,11 +457,18 @@ class SettingsViewController: UITableViewController {
-        let statusView = NetworkProtectionRootView()
-        let hostingController = UIHostingController(rootView: statusView)
-        hostingController.view.backgroundColor = UIColor(Color(designSystemColor: .background))
+        // This will be tidied up as part of https://app.asana.com/0/0/1205084446087078/f
```
I will probably pull this out into a presenter / launcher / router type class.

graeme commented on DuckDuckGoTests/NetworkProtectionInviteViewModelTests.swift:
```diff
@@ -0,0 +1,89 @@
+        XCTAssertEqual(viewModel.currentStep, .codeEntry)
+    }
+
+    // Disabled this test but keeping it around to document the behaviour. It is failing inexplicably.
```
For some reason, errors thrown in the library are not caught in the main app. This does not apply in production, only the test environment. You can manually test this to be sure by entering an incorrect invite code.

diegoreymendez commented on DuckDuckGo.xcodeproj/project.xcworkspace/xcshareddata/swiftpm/Package.resolved:
```diff
@@ -14,9 +14,9 @@
-          "branch": null,
-          "revision": "f255099c2c373d7b0c484849b8903831e7b717ce",
-          "version": "71.0.0"
+          "branch": "graeme/network-protection-test-utils",
```
Friendly visual reminder to help you update the BSK version before merging.

diegoreymendez commented on DuckDuckGo/Assets.xcassets/IntiveLockSuccess.imageset/Contents.json:
```diff
@@ -0,0 +1,12 @@
+{
+  "images" : [
+    {
+      "filename" : "Intive-Lock-Success-96.pdf",
```
Should "Intive" be "Invite"?  This typo is present apparently in several file names, but I'm adding a single note here to avoid overcluttering the PR with comments.

diegoreymendez reviewed (commented):
@graeme - Adding some comments for your consideration.  The implementation looks clean and mostly to be working fine.

Please let me know what you think about my comments, so we can move ahead - if we find agreement in the comments I'd even be happy to tackle it in a follow up if you so prefer.

Additional points:
- The invite code input screen felt like it wasn't working at first and the Submit button remained enabled for a bit allowing me to click it 3 times before it worked.  I'm not sure what's behind this, but I suspect it's network related... it may be a good idea to offer some visual cue that you've submitted (maybe disable the button?).
- I have to long-press the "Clear invite code" button for it to highlight.

graeme committed:
Disable user interaction on making network request

graeme committed:
Fix typo in invite image name

graeme committed:
NetP invite code screen PR 1 (#1881)

Task/Issue URL: https://app.asana.com/0/0/1205084446087083/f
Description:
First PR for the NetP Invite Code flow. You can find designs in the linked task above. This first PR is just to add a root view which will choose between either the (initially empty) Invite or the Status View.

graeme committed:
Merge remote-tracking branch 'origin/develop' into graeme/netp-invite-code-pr-2

graeme committed:
Point to updated BSK release

graeme committed:
Merge remote-tracking branch 'origin/develop' into graeme/netp-invite-code-pr-2
